core/src/apps/cardano/sign_tx.py

```diff
@@ -93,37 +112,62 @@ def _build_inputs(inputs: List[CardanoTxInputType]) -> List[Tuple[bytes, int]]:
 
 
 def _build_outputs(
-    keychain: seed.Keychain, outputs: List[CardanoTxOutputType], protocol_magic: int
+    keychain: seed.Keychain,
+    outputs: List[CardanoTxOutputType],
+    protocol_magic: int,
+    network_id: int,
 ) -> List[Tuple[bytes, int]]:
     result = []
     total_amount = 0
     for output in outputs:
         amount = output.amount
-        if output.address_n:
-            address, _ = derive_address_and_node(
-                keychain, output.address_n, protocol_magic
+        if output.address_parameters:
+            address = derive_address(
+                keychain,
+                output.address_parameters,
+                protocol_magic,
+                network_id,
+                human_readable=False,
             )
         else:
-            address = output.address
-            if address is None:
-                raise wire.ProcessError(
-                    "Each output must have address or address_n field!"
-                )
-            if not is_safe_output_address(address):
+            # unfortunately bytes get decoded as bytearray in protobuf - we need bytes
+            address = bytes(output.address)
+            if not _validate_output_address(address, protocol_magic, network_id):
                 raise wire.ProcessError("Invalid output address!")
 
-            if not matches_with_protocol_magic(address, protocol_magic):
-                raise wire.ProcessError("Output address network mismatch!")
-
         total_amount += amount
-        result.append((base58.decode(address), amount))
+        result.append((address, amount))
 
     if total_amount > LOVELACE_MAX_SUPPLY:
         raise wire.ProcessError("Total transaction amount is out of range!")
 
     return result
 
 
+def _validate_output_address(address: bytes, protocol_magic: int, network_id: int):
+    # todo: GK - which is better?
+    if address is None:
+        # raise wire.ProcessError(
+        #     "Each output must have address or address_n field!"
+        # )
+        return False
+
+    if has_byron_address_header(address):
+        if not is_safe_output_address(address):
+            # raise wire.ProcessError("Invalid output address!")
+            return False
+
+        if not matches_with_protocol_magic(address, protocol_magic):
+            # raise wire.ProcessError("Output address network mismatch!")
+            return False
+    else:
```

Comment:
shouldn't we explicitly check if it has a valid shelley header?

Reply:
Yes, fixed.